File: y2mate.py
import requests
import json
import re


def y2mate(url):
    match = re.search(r"youtube\.com/.*v=([^&]*)", url)
    if match:
        resultIDVideo = match.group(1)
        print(resultIDVideo)
        headers = {
            'accept': "*/*",
            'accept-language': "en-US,en;q=0.9,vi;q=0.8",
            'content-type': "application/x-www-form-urlencoded; charset=UTF-8"
        }
        data = {
            'url': 'https://youtu.be/' + resultIDVideo,
            'q_auto': 0,
            'ajax': 2
        }
        url = 'https://www.y2mate.com/mates/en68/analyze/ajax'
        res = requests.post(url, headers=headers, data=data)
        home = json.loads(res.text)['result']
        imageSrc = home.split('<div class="thumbnail cover">')[
            1].split('<img src="')[1].split('"')[0]
        title = home.split(
            'caption text-left">')[1].split('</div>')[0].split('<b>')[1].replace("</b>", "")
        size = home.split('<a href="#" rel="nofollow">')[
            8].split('<td>')[1].split('</td>')[0]
        type = home.split('<a href="#" rel="nofollow">')[8].split(
            '</a>')[0].split('(')[0].split(".")[1].strip()
        quality = home.split('<a href="#" rel="nofollow">')[
            8].split('</a>')[0].split('(')[1].split(')')[0].split('kbps')[0]
        id = home.split('var k__id =')[1].split(';')[0].split('"')[1].strip()
        urlOne = 'https://www.y2mate.com/mates/en68/convert'
        _headers = {
            'accept': "*/*",
            'accept-language': "en-US,en;q=0.9,vi;q=0.8",
            'content-type': "application/x-www-form-urlencoded; charset=UTF-8"
        }
        _data = {
            'type': 'youtube',
            'v_id': resultIDVideo,
            '_id': id,
            'ajax': '1',
            'token': '',
            'ftype': type,
            'fquality': quality
        }
        res = requests.post(urlOne, headers=_headers, data=_data)
        url = json.loads(res.text)['result']
        urlMusic = url.split('<a href="')[1].split('"')[0]
        infoMusic = {
            'imageSrc': imageSrc,
            'title': title,
            'size': size,
            'type': type,
            'quality': quality,
            'url': urlMusic
        }
        print(json.dumps(infoMusic))
    else:
        result = "Video không hợp lệ"
        return result


y2mate('https://www.youtube.com/watch?v=5zoBkldG1PY')

File: y2mate.js
import axios from "axios";
import { load } from "cheerio";
function y2mate(url) {
    return new Promise(async(resolve, reject) =>{
        if(!await youtube_parser(url)) reject("Không tìm thấy id của video này")
        axios({
            method: 'post',
            url: 'https://www.y2mate.com/mates/en68/analyze/ajax',
            headers: {
                accept: "*/*",
                'accept-language': "en-US,en;q=0.9,vi;q=0.8",
                'content-type': "multipart/form-data"
            },
            data: {
                url: `https://youtu.be/${await youtube_parser(url)}`,
                q_auto: 0,
                ajax: 2
            }
        }).then(async(res) => {
            const $ = load(res.data.result);
            const imageSrc = $('div[class="thumbnail cover"]').find('a > img').attr('src'),
                title = $('div[class="caption text-left"]').find('b').text(),
                size = $('div').find('#mp3 > table > tbody > tr > td:nth-child(2)').text(),
                type = $('div').find('#mp3 > table > tbody > tr > td:nth-child(3) > a').attr('data-ftype'),
                quality = $('div').find('#mp3 > table > tbody > tr > td:nth-child(3) > a').attr('data-fquality'),
                id = /var k__id = "(.*?)"/.exec(res.data.result)[1]
            await axios({
                method: 'post',
                url: 'https://www.y2mate.com/mates/en68/convert',
                headers: {
                    accept: "*/*",
                    'accept-language': "en-US,en;q=0.9,vi;q=0.8",
                    'content-type': "multipart/form-data"
                },
                data:{
                    type: 'youtube',
                    v_id: await youtube_parser(url),
                    _id: id,
                    ajax: '1',
                    token: '',
                    ftype: type,
                    fquality: quality
                }
            }).then(body =>{
                const $ = load(body.data.result);
                const getHref = $('div[class="form-group has-success has-feedback"]').find('a').attr("href");
                resolve({title, size, type, quality, imageSrc, getHref})
            })
        })
    })
}
function youtube_parser(url){
    var regExp = /^.*((youtu.be\/)|(v\/)|(\/u\/\w\/)|(embed\/)|(watch\?))\??v?=?([^#\&\?]*).*/;
    var match = url.match(regExp);
    return (match&&match[7].length==11)? match[7] : false;
}
y2mate()
